x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportDeleteShutdownNodeAction.java

```diff
@@ -77,6 +82,22 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                         .build();
 
                 }
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
```

Comment:
Do we have to do this after the cluster state has been processed? I think I would prefer if possible to do it in a single cluster state update, like with https://github.com/elastic/elasticsearch/blob/eee399aeb5825fc3a98ec62f719ee41d584b4422/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java#L238

Is there a reason we shouldn't inject the `AllocationService` into these transport actions and then if a node was added/removed we can call the reroute from within the `execute` method and only generate a single cluster state update.

What do you think?

Reply:
The advantage of using the reroute service here is that it batches reroutes together. Reroutes are expensive (and idempotent) so batching is preferable even if it costs an extra cluster state update sometimes. If you put/delete a bunch of these shutdown markers all at once then we don't want to do a reroute for each one, and if there are shards starting/failing at the same time then we can piggy-back our reroute on theirs.

I think it'd be better not to notify the listener until after the reroute completes, however: that way we should be able to assert that the shards already started moving when the response comes in.

Reply:
Okay, I didn't realize that reroutes were expensive and that they could be batched, in that case, doing the separate update is better as David said. +1 on waiting until the reroute completes before returning the listener.

Reply:
I originally went this route based on advice that using the reroute service here was better than allocation service, though I didn't know the reasoning why. Thanks Dave!

Updated this to wait until the reroute has been successfully kicked off to respond to the request (after spending way too long trying to figure out why the request kept timing out in the tests, only to realize I forgot the `listener.onResponse(AcknowledgedResponse.TRUE);` in the PUT action, oops).

Reply:
++ I opened https://github.com/elastic/elasticsearch/pull/73512 to record the information about rerouting somewhere easier for others to find.